packages/flutter/lib/src/widgets/binding.dart

```diff
@@ -780,11 +780,14 @@ mixin WidgetsBinding on BindingBase, ServicesBinding, SchedulerBinding, GestureB
       }());
     }
     if (!kReleaseMode) {
-      if (_needToReportFirstFrame && _reportFirstFrame) {
+      if (_needToReportFirstFrame && sendFramesToEngine) {
         developer.Timeline.instantSync('Widgets built first useful frame');
       }
     }
     _needToReportFirstFrame = false;
+    if (firstFrameCallback != null && !sendFramesToEngine) {
+      SchedulerBinding.instance.removeTimingsCallback(firstFrameCallback);
```

Comment:
Is it possible that `sendFramesToEngine = false` here, so `firstFrameCallback` is removed, and then someone calls `allowFirstFrame` without calling `drawFrame` again, and the first frame is rendered without properly completing `_firstFrameCompleter`?

Reply:
allowFirstFrame schedules another frame. To render that frame, drawFrame is called again, which will add that callback again and - if sendFramesToEngine is now true - will complete _firstFrameCompleter.

Reply:
I see. I guess it's due to the following code?
```
if (!_firstFrameSent)
      scheduleWarmUpFrame();
```